src/config/mod.rs

```diff
@@ -0,0 +1,60 @@
+use std::any::{Any, TypeId};
+use std::sync::Arc;
+use std::collections::HashMap;
+use std::hash::{BuildHasherDefault, Hasher};
+
+type AnyMap = HashMap<TypeId, Box<Any + Send + Sync>, BuildHasherDefault<IdHasher>>;
+
+#[derive(Debug, Default)]
+struct IdHasher(u64);
+
+impl Hasher for IdHasher {
+    fn write(&mut self, _: &[u8]) {
+        unreachable!("TypeId calls write_u64");
+    }
+
+    #[inline]
+    fn write_u64(&mut self, id: u64) {
+        self.0 = id;
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        self.0
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ConfigBuilder {
+    inner: AnyMap,
+}
+
+impl ConfigBuilder {
+    pub(crate) fn new() -> Self {
+        Self { inner: AnyMap::default() }
+    }
+
+    pub(crate) fn insert<T: Send + Sync + 'static>(mut self, val: T) -> Self {
+        self.inner.insert(TypeId::of::<T>(), Box::new(val));
+        self
+    }
+
+    pub(crate) fn into_config(self) -> Config {
+        Config { inner: Arc::new(self.inner) }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Config {
+    inner: Arc<AnyMap>,
+}
+
+impl Config {
+    pub fn get<T: Send + Sync + 'static>(&self) -> Option<&T> {
```

Comment:
I'm not 100% sure, but I don't think we need to include all thee bounds here, just enough to call `TypeId::of`.